### ISSUE TITLE
[BUG] Node auto upgrade not working correctly

### ISSUE THREAD

**Issue by stevehipwell** [Azure/AKS#4115]
**Describe the bug**
We're experiencing an failure to upgrade nodes to `AKSUbuntu-2204gen2containerd-202402.07.0` on multiple AKS clusters set to use the node image channel. Some of these clusters have been set to weekly upgrades and some have been set to daily upgrades. The upgrades are available manually in the console.

For example on an AKS cluster in westeurope which had it's node auto upgrade maintenance window configured in December 2023 to a daily node image upgrade we did not get the `AKSUbuntu-2204gen2containerd-202402.07.0` when it was released and there were no warning or errors recorded. I managed to get this cluster to upgrade last night after changing the window in the console and then re-applying the desired configuration (see other issue for why this couldn't be done in the console).

**To Reproduce**
I'm not sure we would be able to reproduce this.

**Expected behavior**
I'd expect that an AKS cluster with a daily node image maintenance window would upgrade to the latest node image when it was available. If the upgrade couldn't be applied I'd expect to see an error or a warning.

**Screenshots**
I don't have these.

**Environment (please complete the following information):**
 -  CLI Version [e.g. 3.22]: n/a
 - Kubernetes version [e.g. 1.24.3]: 1.28.3
 - CLI Extension version [e.g. 1.7.5] if applicable: n/a
 - Browser [e.g. chrome, safari] is applicable: n/a

**Additional context**

**Comment by stevehipwell** [comment]
@philwelz I've replied on #4116. I think we need to get to the bottom of which documentation is correct and if daily isn't supported why the CLI supports it, the console shows it as the configured settings, and why it's been working up until now.

**Comment by philwelz** [comment]
@stevehipwell you are running daily maintenance on the security channel type `Node Image` that supports currently only weekly checks. So only if the weekly check shows changes the daily maintenance `aksManagedNodeOSUpgradeSchedule`  is executed. 

You can see it from the docs -> `aksManagedNodeOSUpgradeSchedule` controls when the node operating system security patching scheduled by your node OS auto-upgrade channel are performed.

So i come back to the rollout order, westeurope is rlly late on the rollout order. Im guessing as you re-applied the desired configuration it initiated a check and found a new node os image. I would go and verified it when the next image is releases.

**Comment by stevehipwell** [comment]
@philwelz we're setting the `aksManagedNodeOSUpgradeSchedule` which according to the [documentation](https://learn.microsoft.com/en-us/azure/aks/planned-maintenance#daily-schedule) supports a daily maintenance window; and the CLI validation supports this. Are you saying that this will run daily but will only do something based on a seperate weekly schedule?

If so firstly this isn't clear in the documentation, secondly it doesn't give us the option for a strong security posture, and finally it doesn't seem to be deterministic (an incidental change would seem to cause a recalculation).

**Comment by kaarthis** [comment]
Yes typically we ship images weekly cadence so like you said if they are available in westeurope during the window start it will upgrade else do nothing. However i do agree we need better notifications of a no image window / skip - We are actually working on an AKS communication manager for this summer : that provides these type of notices and also for Auto upgrade failures.

**Comment by stevehipwell** [comment]
@kaarthis the issue I'm reporting is that the window (daily) was evaluated at least twice while there was an available image upgrade and it didn't get applied.

Based on the current docs my understanding of how node image upgrades work is as follows.

- A new image gets released weekly
- This image is rolled out to the regions
- A daily node upgrade window should upgrade the nodes if a new image is available in the region

**Comment by stevehipwell** [comment]
@kaarthis the window configuration is below. The availability of new images was confirmed by the AKS release tracker and the console.

```shell
~ > az aks maintenanceconfiguration list --resource-group xxxxxxxx-aks-8 --cluster-name aks-8
The behavior of this command has been altered by the following extension: aks-preview
[
  {
    "id": "/subscriptions/xxxxxxxx/resourceGroups/xxxxxxxx-aks-8/providers/Microsoft.ContainerService/managedClusters/aks-8/maintenanceConfigurations/aksManagedAutoUpgradeSchedule",
    "maintenanceWindow": {
      "durationHours": 4,
      "notAllowedDates": null,
      "schedule": {
        "absoluteMonthly": null,
        "daily": null,
        "relativeMonthly": null,
        "weekly": {
          "dayOfWeek": "Sunday",
          "intervalWeeks": 1
        }
      },
      "startDate": "2023-12-14",
      "startTime": "00:00",
      "utcOffset": "+00:00"
    },
    "name": "aksManagedAutoUpgradeSchedule",
    "notAllowedTime": null,
    "resourceGroup": "xxxxxxxx-aks-8",
    "systemData": null,
    "timeInWeek": null,
    "type": null
  },
  {
    "id": "/subscriptions/xxxxxxxx/resourceGroups/xxxxxxxx-aks-8/providers/Microsoft.ContainerService/managedClusters/aks-8/maintenanceConfigurations/aksManagedNodeOSUpgradeSchedule",
    "maintenanceWindow": {
      "durationHours": 4,
      "notAllowedDates": null,
      "schedule": {
        "absoluteMonthly": null,
        "daily": {
          "intervalDays": 1
        },
        "relativeMonthly": null,
        "weekly": null
      },
      "startDate": "2024-02-24",
      "startTime": "00:00",
      "utcOffset": "+00:00"
    },
    "name": "aksManagedNodeOSUpgradeSchedule",
    "notAllowedTime": null,
    "resourceGroup": "xxxxxxxx-aks-8",
    "systemData": null,
    "timeInWeek": null,
    "type": null
  }
]
```

**Comment by kaarthis** [comment]
Hmm it appears you are using the correct configuration and setting to 'Daily' NodeImage schedule as expected. The only thing that pops up is the start date of 2/24 may have been too close for the sunday image to be applied - As in we call out : https://learn.microsoft.com/en-us/azure/aks/planned-maintenance?tabs=azure-cli#faq

_I configured a maintenance window, but upgrade didn't happen - why?

AKS auto-upgrade needs a certain amount of time to take the maintenance window into consideration. We recommend at least 24 hours between the creation or update of a maintenance configuration the scheduled start time._

Please check if this could be the case . Also did this work after wards?

**Comment by stevehipwell** [comment]
@kaarthis the maintenance window was definitely configured more than 24h before the image upgrade was released.

We're seeing similar (or the same) issues again with nodes on very outdated image versions despite being set to check for new versions daily.
